Integrate QuicTransport server with other wpt tools

Hexcles opened the pull request:
Now there are four ways to start the QUIC server, each one delegating to the previous layer:

1. (Standalone mode) Run `python3 tools/quic/serve.py` directly. This isn't really useful as you'd need to install the dependencies yourself.
2. `wpt serve-quic-transport` brings up only the QUIC server, but using the `wpt` frontend to manage virtualenv automatically (`_venv3`). This is mostly for internal plumbing only.
3. `wpt serve --quic-transport` brings up `wptserve` as well as `wpt serve-quic-transport`. This is the main entry point. `--quic-transport` won't be needed once the QUIC server is enabled by default.
4. `wpt run --enable-quic ...` can run tests that require QUIC servers (currently just the QUIC Transport server) and essentially uses `wpt serve --quic-transport` under the hood.

Hexcles committed:
Get "wpt serve-quic" working

Hexcles committed:
Some stylistic improvements

Hexcles committed:
[serve] Add QUIC server to "wpt serve"

Start `wpt serve-quic` if the "quic" port is specified in config.json
(e.g. plumbed from wptrunner when `wpt run --enable-quic`), or the
"--quic" flag is given in CLI.

Hexcles committed:
fix tests

Hexcles commented:
@yutakahirano I've got almost everything up and running now, except this error from aioquic:

```
INFO:root:QUIC event: <class 'aioquic.quic.events.ProtocolNegotiated'>                                                                                                                                                                                                                                                   
INFO:quic:[fd2aec89dd8db3dc] Connection close code 0xA, reason 28:TLS handshake failure (ENCRYPTION_HANDSHAKE) 46: certificate unknown                                                                                                                                                         
INFO:root:QUIC event: <class 'aioquic.quic.events.ConnectionTerminated'>
```

(I'm using `-c tools/quic/certs/cert.pem -k tools/quic/certs/cert.key`.)

yutakahirano commented:
Sorry for the late reply.

@Hexcles did you give appropriate commandline parameters to Chromium? [1] maybe helpful. Also, recent Chrome canary prints an error message to the console log on handshake error. Can you tell me what it says?

1: https://docs.google.com/document/d/1vnv4kwiGesCAaAjFlXR1XZHN3YeHe9xZcTmIvSoX1aU/edit#heading=h.ru2qqp8h0kx2

yutakahirano commented:
I have one question. Which do you prefer, "quic" or "quic-transport"? QuicTransport(https://tools.ietf.org/html/draft-vvv-webtransport-quic) is a protocol built on top of QUIC so QUIC may be misleading.

My original intention was having QUIC related tools (including QuicTransport related ones) under tools/quic. webtransport/quic is not a problem because "QUIC" in the context of WebTransport surely means QuicTransport.

But I'm not sure unqualified "QUIC" (e.g., wpt serve-quic) is good.

Hexcles committed:
Merge branch 'master' into start-quic

Hexcles committed:
Rename serve-quic to serve-quic-transport

Also rename the name of the port from "quic" to "quic-transport" for
clarity.

This is to clarify that this port runs a QUIC transport server instead
of e.g. a HTTP/3 server.

Hexcles committed:
Bump aioquic to 0.8.8

Hexcles committed:
Improve error logging for quic_transport_server.py

Hexcles committed:
Fix handlers_path

Hexcles commented on tools/quic/quic_transport_server.py:
```diff
@@ -167,73 +167,29 @@ def pop(self, label: bytes) -> Optional[SessionTicket]:
-    handlers_path = os.path.abspath(os.path.expanduser(args.handlers_path))
-    logging.log(logging.INFO, 'port = %s' % args.port)
-    logging.log(logging.INFO, 'handlers path = %s' % handlers_path)
+    global handlers_path
```
Missing this was a bug that caused "Client indication error: unsupported operand type(s) for +: 'NoneType' and 'str'".

yutakahirano replied:
Yeah, previously this code was in the global scope but now it's in a function.

Hexcles commented on tools/serve/serve.py:
```diff
@@ -702,6 +704,63 @@ def start_wss_server(host, port, paths, routes, bind_address, config, **kwargs):
+        return QuicTransportDaemon(host,
+                          port,
+                          private_key=config.ssl_config["key_path"],
+                          certificate=config.ssl_config["cert_path"],
```
Note that this means when running `wpt serve --quic-transport` we use the certs from `tools/certs` by default, which now works with aioquic 0.8.8.

Hexcles reviewed (commented):
@yutakahirano Thanks for the link. I didn't know "--ignore-certificate-errors does not work for this feature". Explicitly specifying SPKI works.

Now with the right flags, Chrome 84 Dev passes `webtransport/quic/client-indication.sub.any.html`. Server was started with `./wpt serve --quic-transport`.

PTAL again. I've highlighted a few critical changes.

yutakahirano commented on tools/quic/quic_transport_server.py:
```diff
@@ -16,8 +15,9 @@
+logger = logging.getLogger(__name__)
 
-handlers_path = None
+handlers_path = ""
```
Single quotations are used in this file so doing so here is probably better for consistency.

Hexcles replied:
```suggestion
handlers_path = ''
```

yutakahirano reviewed (approved):
Thank you, LGTM!

Hexcles committed:
Nit
